### PR TITLE
L2Sender reinit fix

### PR DESCRIPTION
**Pull request by maurelian**
Replaces #8797 which was accidentally closed.

**Description**

Fixes an issue which would allow for a single messages to be replayed twice by intercepting an upgrade.

This is addressed in two ways: 

- Do not over the `xDomainMsgSender` if it is not the default value. 
- Added another check that `successfulMessages` for the message is still false after the
external call. 

**Exploit Description**

To perform the exploit, the following steps must be taken by the attacker:

1. Wait for a signed upgrade transaction.
2. Once the signed upgrade transaction is available, front-run by running it inside a
withdrawal transaction.
3. The attacker's withdrawal payload will call their own contract which would run the
upgrade transaction and then re-enter `relayMessage()` with their own same
withdrawal message.
4. The reentrancy will succeed since `xDomainMsgSender` has been reset by the upgrade.

The attacker must also satisfy the following with their withdrawal message:

1. It must be a failed withdrawal so they can re-enter `relayMessage()`. A fresh withdrawal
cannot reenter as it checks `failedMessages[versionedHash]` is true.
2. It must have a value set. The amount for value is the amount that they will able to
drain from the contract.

### CODE REVIEW

File: packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol [hunk 1]
```diff
@@ -631,6 +632,8 @@ contract L1CrossDomainMessenger_ReinitReentryTest is Bridge_Initializer {
             l1CrossDomainMessenger.initialize(SuperchainConfig(superchainConfig));
 
             // attempt to re-replay the withdrawal
+            vm.expectEmit(address(l1CrossDomainMessenger));
+            emit FailedRelayedMessage(hash);
```
Comment (maurelian) [inline packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol:635]: Addresses [this request](https://github.com/ethereum-optimism/optimism/pull/8797#discussion_r1442353495): 
> Could we add an expect emit here to clarify what codepath this is supposed to take?

File: packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol [hunk 2]
```diff
@@ -288,6 +288,12 @@ abstract contract CrossDomainMessenger is
         xDomainMsgSender = Constants.DEFAULT_L2_SENDER;
 
         if (success) {
+            // This check is identical to one above, but it ensures that the same message cannot be relayed
+            // twice, and adds a layer of protection against rentrancy.
+            require(
+                successfulMessages[versionedHash] == false,
+                "CrossDomainMessenger: message has already been relayed via reentrancy"
+            );
```
Comment (maurelian) [inline packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol]: A [comment](https://github.com/ethereum-optimism/optimism/pull/8797#discussion_r1442352945) was left on this line in the other PR: 
> I don't see any expect reverts for this case

Comment (maurelian) [reply packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol]: This line should be unreachable. 

Prior to the other fix in this contract (on the initializer), this line would only be reachable during the reinit intercept attack. That makes this change redundant, but IMO it is a good idea as it implements a more FREI-PI like pattern that enforces a pretty basic invariant that no message can be relayed twice, which feels really good to be enforcing. 

However I agree that it's bad practice to have an error message that we can't test. Therefore I will follow the pattern we've sent elsewhere in this contract of using an `assert()` for unreachable conditions:

https://github.com/ethereum-optimism/optimism/blob/jm/fix-l2sender-reinit/packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol#L244-L245

File: packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol [hunk 3]
```diff
@@ -4,6 +4,7 @@ pragma solidity 0.8.15;
 // Testing utilities
 import { Bridge_Initializer } from "test/setup/Bridge_Initializer.sol";
 import { Reverter, ConfigurableCaller } from "test/mocks/Callers.sol";
+import { console2 as console } from "forge-std/console2.sol";
```
Comment (tynes) [inline packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol]: dead import

Comment (maurelian) [reply packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol]:
```suggestion
```

**Review by tynes — APPROVED** [review]

File: packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol [hunk 4]
```diff
@@ -354,7 +357,13 @@ abstract contract CrossDomainMessenger is
     /// @notice Initializer.
     // solhint-disable-next-line func-name-mixedcase
     function __CrossDomainMessenger_init() internal onlyInitializing {
-        xDomainMsgSender = Constants.DEFAULT_L2_SENDER;
+        // We only want to set the xDomainMsgSender to the default value if it hasn't been initialized yet,
+        // meaning that this is a fresh contract deployment.
+        // This prevents resetting the xDomainMsgSender to the default value during an upgrade, which would enable
+        // a reentrant withdrawal to sandwhich the upgrade replay a withdrawal twice.
```
Comment (mds1) [inline packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:363]: nit 😅
```suggestion
        // a reentrant withdrawal to sandwich the upgrade to replay a withdrawal twice.
```

**Pull request merged**